### ISSUE TITLE
[in_app_purchase][iOS] Uncompleted PurchaseDetail items have applicationUsername set to NULL following restart

### ISSUE THREAD

**Issue by jesper372** [flutter/flutter#129203]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

On iOS:

1. Initiate a purchase -- passing in a UUID to the `applicationUserName` argument of the `PurchaseParam` object -- but force the verification of the purchase to fail, so that it is not completed. Observe that when the `PurchaseDetails` item appears in the stream, its `applicationUsername` field (available via `(purchaseDetails as AppStorePurchaseDetails).skPaymentTransaction.payment.applicationUsername`) contains the UUID that was passed to the `PurchaseParam` object, as expected.
2. Restart the app. A `PurchaseDetails` item appears in the stream, for the uncompleted purchase. However, its `applicationUsername` field no longer contains the UUID that was passed to the `PurchaseParam` object; it is now `null`.

[The documentation](https://developer.apple.com/documentation/storekit/skmutablepayment/1506088-applicationusername?language=objc) advocates the use of a UUID for `applicationUsername`, and suggests that as long as a UUID _is_ used, it will persist.

Note that, unlike in https://github.com/flutter/flutter/issues/109672, there is no explicit querying for restored purchases in the above sequence (i.e. in that issue, `restorePurchases` is called, and a specific `applicationUserName` argument is passed to it).

### Expected results

Following app restart, the `PurchaseDetails` item that appears in the stream should have its `applicationUsername` field set to the UUID that was used when initiating the purchase (as the `PurchaseDetails` item that appeared prior to the restart did).

### Actual results

Following app restart, the `PurchaseDetails` item that appears in the stream has its `applicationUsername` field set to `null`.

### Code sample

```
import 'dart:async';
import 'dart:io';

import 'package:flutter/material.dart';
import 'package:in_app_purchase/in_app_purchase.dart';
import 'package:in_app_purchase_android/in_app_purchase_android.dart';
import 'package:in_app_purchase_storekit/in_app_purchase_storekit.dart';

void main() {
  WidgetsFlutterBinding.ensureInitialized();
  runApp(const MyApp());
}

const String _cons1y = 'cons1y';
const List<String> _productIds = <String>[_cons1y];

class MyApp extends StatefulWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  final InAppPurchase _inAppPurchase = InAppPurchase.instance;
  late StreamSubscription<List<PurchaseDetails>> _subscription;
  List<ProductDetails> _products = <ProductDetails>[];
  bool _isAvailable = false;
  bool _purchasePending = false;
  bool _loading = true;
  String? _queryProductError;
  final uuid = '1293a822-417d-4edb-9b7c-4ee406b7b7d8';

  @override
  void initState() {
    super.initState();
    final Stream<List<PurchaseDetails>> purchaseStream = _inAppPurchase.purchaseStream;
    _subscription = purchaseStream.listen((List<PurchaseDetails> purchaseDetailsList) {
      _processPurchases(purchaseDetailsList);
    }, onDone: () {
      _subscription.cancel();
    }, onError: (Object error) {
    });
    initStoreInfo();
  }

  Future<void> initStoreInfo() async {
    final bool isAvailable = await _inAppPurchase.isAvailable();
    if (!isAvailable) {
      setState(() {
        _isAvailable = isAvailable;
        _products = <ProductDetails>[];
        _purchasePending = false;
        _loading = false;
      });
      return;
    }

    final ProductDetailsResponse productDetailResponse = await _inAppPurchase.queryProductDetails(_productIds.toSet());
    if (productDetailResponse.error != null) {
      setState(() {
        _queryProductError = productDetailResponse.error!.message;
        _isAvailable = isAvailable;
        _products = productDetailResponse.productDetails;
        _purchasePending = false;
        _loading = false;
      });
      return;
    }

    if (productDetailResponse.productDetails.isEmpty) {
      setState(() {
        _queryProductError = null;
        _isAvailable = isAvailable;
        _products = productDetailResponse.productDetails;
        _purchasePending = false;
        _loading = false;
      });
      return;
    }

    setState(() {
      _isAvailable = isAvailable;
      _products = productDetailResponse.productDetails;
      _purchasePending = false;
      _loading = false;
    });
  }

  @override
  void dispose() {
    _subscription.cancel();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    final List<Widget> stack = <Widget>[];
    if (_queryProductError == null) {
      stack.add(
        ListView(
          children: <Widget>[
            _buildProductList(),
          ],
        ),
      );
    } else {
      stack.add(Center(
        child: Text(_queryProductError!),
      ));
    }
    if (_purchasePending) {
      stack.add(
        const Stack(
          children: <Widget>[
            Opacity(
              opacity: 0.3,
              child: ModalBarrier(dismissible: false, color: Colors.grey),
            ),
            Center(
              child: CircularProgressIndicator(),
            ),
          ],
        ),
      );
    }

    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('IAP Example'),
        ),
        body: Stack(
          children: stack,
        ),
      ),
    );
  }

  Card _buildProductList() {
    if (_loading) {
      return const Card(child: ListTile(leading: CircularProgressIndicator(), title: Text('Fetching products...')));
    }
    if (!_isAvailable) {
      return const Card();
    }
    final List<ListTile> productList = <ListTile>[];
    productList.addAll(_products.map(
      (ProductDetails productDetails) {
        return ListTile(
          title: Text(productDetails.title),
          subtitle: Text(productDetails.description),
          trailing: TextButton(
            style: TextButton.styleFrom(foregroundColor: Colors.white, backgroundColor: Colors.green[800]),
            onPressed: () {
              if (Platform.isAndroid) {
                _inAppPurchase.buyConsumable(
                    purchaseParam: GooglePlayPurchaseParam(
                      productDetails: productDetails,
                      applicationUserName: uuid,
                    )
                );
              } else {
                _inAppPurchase.buyNonConsumable(
                  purchaseParam: PurchaseParam(
                    productDetails: productDetails,
                    applicationUserName: uuid,
                  ),
                );
              }
            },
            child: Text(productDetails.price),
          ),
        );
      },
    ));

    return Card(child: Column(children: productList));
  }

  void _showPendingUI() {
    setState(() {
      _purchasePending = true;
    });
  }

  void _handlePendingPurchaseCancelled(PurchaseDetails purchaseDetails) {
    setState(() {
      _purchasePending = false;
    });
  }

  Future<void> _deliverProduct(PurchaseDetails purchaseDetails) async {
    setState(() {
      _purchasePending = false;
    });
  }

  void _handleError(IAPError error) {
    setState(() {
      _purchasePending = false;
    });
  }

  Future<bool> _verifyPurchase(PurchaseDetails purchaseDetails) async {
    // Simulate verification of the purchase failing (say, because the backend is not reachable):
    return false;
  }

  void _handleInvalidPurchase(PurchaseDetails purchaseDetails) {
    setState(() {
      _purchasePending = false;
    });
  }

  Future<void> _processPurchases(List<PurchaseDetails> purchaseDetailsList) async {
    for (final PurchaseDetails purchaseDetails in purchaseDetailsList) {
      if (purchaseDetails is AppStorePurchaseDetails) {
        if (purchaseDetails.skPaymentTransaction.payment.applicationUsername != uuid) {
          // This is the problem: we want to ignore the uncompleted purchase item if it didn't originate from the
          // currently signed-in user (i.e. the user with id `uuid`), but following app restart, the purchase item's
          // `applicationUsername` field is `null`, thus making it impossible to know who attempted the purchase.
          // Note that in the purchase item that appears here prior to the restart, the `applicationUsername` field
          // *does* match `uuid`.
          continue;
        }
      }
      if (purchaseDetails.status == PurchaseStatus.pending) {
        _showPendingUI();
      } else if (purchaseDetails.status == PurchaseStatus.canceled) {
        _handlePendingPurchaseCancelled(purchaseDetails);
      } else {
        if (purchaseDetails.status == PurchaseStatus.error) {
          _handleError(purchaseDetails.error!);
        } else if (purchaseDetails.status == PurchaseStatus.purchased) {
          final bool valid = await _verifyPurchase(purchaseDetails);
          if (valid) {
            _deliverProduct(purchaseDetails);
          } else {
            _handleInvalidPurchase(purchaseDetails);
            return;
          }
        }
        if (purchaseDetails.pendingCompletePurchase) {
          await _inAppPurchase.completePurchase(purchaseDetails);
        }
      }
    }
  }
}
```

### Flutter Doctor output

```
[✓] Flutter (Channel stable, 3.10.3, on macOS 12.6.6 21G646 darwin-x64, locale en-GB)
    • Flutter version 3.10.3 on channel stable at /Users/jesper/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision f92f44110e (3 weeks ago), 2023-06-01 18:17:33 -0500
    • Engine revision 2a3401c9bb
    • Dart version 3.0.3
    • DevTools version 2.23.1

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at /Users/jesper/Library/Android/sdk
    • Platform android-33, build-tools 33.0.0
    • ANDROID_HOME = /Users/jesper/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.2)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14C18
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)

[✓] VS Code (version 1.78.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension can be installed from:
      🔨 https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-x64     • macOS 12.6.6 21G646 darwin-x64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 114.0.5735.133

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

**Comment by darshankawar** [comment]
Thanks for the detailed report and code sample. Using it in local project and `applicationUserName` being set as UUID and upon restarting the app, I observed the same as mentioned:

`applicationUsername field does not contain the UUID that was passed to the PurchaseParam object; it is null.`

<details>
<summary> stable, master flutter doctor -v </summary>

```
[!] Flutter (Channel stable, 3.10.5, on macOS 12.2.1 21D62 darwin-x64, locale
    en-GB)
    • Flutter version 3.10.5 on channel stable at
      /Users/dhs/documents/fluttersdk/flutter
    ! Warning: `flutter` on your path resolves to
      /Users/dhs/Documents/Fluttersdk/flutter/bin/flutter, which is not inside
      your current Flutter SDK checkout at
      /Users/dhs/documents/fluttersdk/flutter. Consider adding
      /Users/dhs/documents/fluttersdk/flutter/bin to the front of your path.
    ! Warning: `dart` on your path resolves to
      /Users/dhs/Documents/Fluttersdk/flutter/bin/dart, which is not inside your
      current Flutter SDK checkout at /Users/dhs/documents/fluttersdk/flutter.
      Consider adding /Users/dhs/documents/fluttersdk/flutter/bin to the front
      of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 796c8ef792 (2 days ago), 2023-06-13 15:51:02 -0700
    • Engine revision 45f6e00911
    • Dart version 3.0.5
    • DevTools version 2.23.1
    • If those were intentional, you can disregard the above warnings; however
      it is recommended to use "git" directly to perform update checks and
      upgrades.

[!] Xcode - develop for iOS and macOS (Xcode 12.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    ! Flutter recommends a minimum Xcode version of 13.
      Download the latest version or update via the Mac App Store.
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] VS Code (version 1.62.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (5 available)
    • SM G975F (mobile)       • RZ8M802WY0X • android-arm64   • Android 11 (API 30)
    • Darshan's iphone (mobile)  • 21150b119064aecc249dfcfe05e259197461ce23 •
      ios            • iOS 14.4.1 18D61
    • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729     •
      ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)            • macos                                    •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)               • chrome                                   •
      web-javascript • Google Chrome 98.0.4758.80

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.

[!] Flutter (Channel master, 3.12.0-7.0.pre.42, on macOS 12.2.1 21D62
    darwin-x64, locale en-GB)
    • Flutter version 3.12.0-7.0.pre.42 on channel master at
      /Users/dhs/documents/fluttersdk/flutter
    ! Warning: `flutter` on your path resolves to
      /Users/dhs/Documents/Fluttersdk/flutter/bin/flutter, which is not inside
      your current Flutter SDK checkout at
      /Users/dhs/documents/fluttersdk/flutter. Consider adding
      /Users/dhs/documents/fluttersdk/flutter/bin to the front of your path.
    ! Warning: `dart` on your path resolves to
      /Users/dhs/Documents/Fluttersdk/flutter/bin/dart, which is not inside your
      current Flutter SDK checkout at /Users/dhs/documents/fluttersdk/flutter.
      Consider adding /Users/dhs/documents/fluttersdk/flutter/bin to the front
      of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 04ff86f020 (2 hours ago), 2023-06-20 19:40:48 -0700
    • Engine revision e5a860c547
    • Dart version 3.1.0 (build 3.1.0-234.0.dev)
    • DevTools version 2.24.0
    • If those were intentional, you can disregard the above warnings; however
      it is recommended to use "git" directly to perform update checks and
      upgrades.

[!] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Users/dhs/Library/Android/sdk
    ✗ cmdline-tools component is missing
      Run `path/to/sdkmanager --install "cmdline-tools;latest"`
      See https://developer.android.com/studio/command-line for more details.
    ✗ Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/macos#android-setup for
      more details.

[✓] Xcode - develop for iOS and macOS (Xcode 13.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 13C100
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] IntelliJ IDEA Ultimate Edition (version 2021.3.2)
    • IntelliJ at /Applications/IntelliJ IDEA.app
    • Flutter plugin version 65.1.4
    • Dart plugin version 213.7228

[✓] VS Code (version 1.62.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.29.0

[✓] Connected device (3 available)
    • Darshan's iphone (mobile) • 21150b119064aecc249dfcfe05e259197461ce23 • ios
      • iOS 15.3.1 19D52
    • macOS (desktop)           • macos                                    •
      darwin-x64     • macOS 12.2.1 21D62 darwin-x64
    • Chrome (web)              • chrome                                   •
      web-javascript • Google Chrome 109.0.5414.119

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
      
[!] Xcode - develop for iOS and macOS (Xcode 12.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    ! Flutter recommends a minimum Xcode version of 13.
      Download the latest version or update via the Mac App Store.
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] VS Code (version 1.62.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (5 available)
    • SM G975F (mobile)       • RZ8M802WY0X • android-arm64   • Android 11 (API 30)
    • Darshan's iphone (mobile)  • 21150b119064aecc249dfcfe05e259197461ce23 •
      ios            • iOS 14.4.1 18D61
    • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729     •
      ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)            • macos                                    •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)               • chrome                                   •
      web-javascript • Google Chrome 98.0.4758.80

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.



```

</details>

**Comment by tienbm** [comment]
same issue

**Comment by stuartmorgan** [comment]
@mvanbeusekom Is this one you could take a look at?

**Comment by mvanbeusekom** [comment]
@stuartmorgan yes this is something we can look at.

**Comment by imsujan276** [comment]
I am facing the same issue. Previously, the callback used to contain the applicationUserName as authName but after the update or later, the callback does not include it. It is making it hard to validate the user assigned to the purchase

**Comment by tienbm** [comment]
Yes. hope will fix soon

**Comment by nduc86686** [comment]
hope you guys will solve it soon, it's been almost 1 year 🥇

**Comment by mvanbeusekom** [comment]
@stuartmorgan, did some additional debugging to see if this is an error in the plugin. I have put a breakpoint on the native implementation of the `paymentQueue:updatedTransactions:` method in the plugin and scanned the information received from the StoreKit framework. As per my suspicions the original transaction details also don't provide the `applicationUsername` value. The screenshot below show the details:

![Image](https://github.com/flutter/flutter/assets/13098888/a1f5189c-6bbe-4a1d-a9d5-cbc03ddb6fea)

I have followed the instructions from the original poster and can confirm that with the initial verification the `applicationUsername` is provided, however when you restart the application and Apple sends the transaction again the `applicationUsername` parameter is not specified as part of the transaction details.

In short this isn't a bug in the in_app_purchase plugin, but simple how the StoreKit SDK behaves. 

Regarding the documentation mentioned, Apple does explain the `applicationUsername` is persisted (provided it is in the form of an UUID value), however they only mention that it will be delivered via the App Store Server API, App Store Server Notifications or through the `verifyReceipt` method. These are all related to methods that should be handled by the applications backend system. There don't seem to be any specific references to the app specific APIs.
